ENH: xz compression_args df.to_pickle(path, compression={'method':'xz','pr…

segatrade opened the pull request:
…eset': level})
level = 9

- [x] closes #52979 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

segatrade committed:
xz compression_args df.to_pickle(path, compression={'method':'xz','preset': 9})

segatrade committed:
fix Line too long

pandas/io/common.py:829:89: E501 Line too long (103 > 88 characters)

segatrade committed:
fix black

twoertwein commented:
Thank you for the PR. You definitely will need the following:
- tests
- update the documentation https://github.com/pandas-dev/pandas/blob/8cf4ab491b06274f555d1ea305c172354208747a/pandas/core/shared_docs.py#L462 (and for decompression below that)
- and a whatsnew entry: doc/source/whatsnew/v2.1.0.rst

segatrade committed:
xz compression_args docs

segatrade committed:
ruff 458:89: E501 Line too long

segatrade committed:
fix mypy Failed
pandas/io/common.py:830: error: Argument 1 to "LZMAFile" has incompatible type "Union[str, BaseBuffer]"; expected "Optional[Union[Union[str, bytes, PathLike[str], PathLike[bytes]], IO[bytes]]]"  [arg-type]
pandas/io/common.py:831: error: Unused "type: ignore" comment

twoertwein commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -100,7 +100,7 @@ Other enhancements
 - Performance improvement in :func:`concat` with homogeneous ``np.float64`` or ``np.float32`` dtypes (:issue:`52685`)
 - Performance improvement in :meth:`DataFrame.filter` when ``items`` is given (:issue:`52941`)
--
+- Added ``**compression_args`` parameter to :meth:`DataFrame.to_pickle` (:issue:`52979`) for preset=level
```
Maybe something along the lines of

> - Many read/to_* functions, such as :meth:`DataFrame.to_pickle` and :func:`read_csv`, support forwarding compression arguments to `lzma.LZMAFile` (:issue:`52979`)

twoertwein commented:
> Thank you for advices and maintaining pandas! Seems done with checks and docs. Do I need add any new tests or it's ready?

Yes, you will need a test. Probably in https://github.com/pandas-dev/pandas/blob/main/pandas/tests/io/test_compression.py I'm not sure how to best test that the keywords are being forwarded: maybe one test with either to_pickle/read_csv and just pass a valid argument and then another test that passes an invalid keyword that raises an error in lzma to ensure the keyword is forwarded to it.

segatrade committed:
xz compression tests
docs ench update

Commit:
Sphinx lint & black fix

Commit:
additional assert in test
sort whatsnew entries alphabetically & black fix

Commit:
fix test obj

Commit:
fix path test

twoertwein commented on pandas/tests/io/test_compression.py:
```diff
@@ -252,6 +252,24 @@ def test_gzip_compression_level(obj, method):
+        compressed_size_default = os.path.getsize(path)
+        getattr(obj, method)(path, compression={"method": "xz", "preset": 1})
+        compressed_size_fast = os.path.getsize(path)
+        assert compressed_size_default < compressed_size_fast
```
Great that you found a way to test the forwarded arguments directly! In that case, you can probably remove the `test_xz_compression_invalid_args`

Commit:
removed test_xz_compression_invalid_args , fix read_csv(compression

segatrade commented:
Thank you! Done. Seems now everything is ready for review.

mroeschke commented:
Thanks @segatrade
